### ISSUE TITLE
Inconsistent MethodSignatureMismatch reporting: it yields randomly and poiting to the wrong class

### ISSUE THREAD

**Issue by fluffycondor** [vimeo/psalm#6773]
psalm 4.11.2

Sometimes Psalm yields MethodSignatureMismatch errors on some of my project's Repositories classes extended from the Doctrine one.

> ERROR: MethodSignatureMismatch - src/Repository/FooRepository.php:15:7 - Argument 1 of Doctrine\ORM\EntityRepository::findBy has wrong type 'array<array-key, mixed>', expecting 'array<string, mixed>' as defined by Doctrine\Persistence\ObjectRepository::findBy (see https://psalm.dev/042)
> class FooRepository extends ServiceEntityRepository

If I do `psalm --clear-cache` then this kind of errors disappears for some time. (?!)

One more interesting thing is that Doctrine's code is perfectly correct, the first argument is typed array<string, mixed> in both cases (it's a recent change, it was just an array<mixed>). But! There are `@psalm-method` stubs in the FooRepository itself that actually causes the issue:
````php
/*
 * @psalm-method list<Foo> findBy(array $criteria, array $orderBy = null, $limit = null, $offset = null)
 */
class FooRepository extends ServiceEntityRepository
{
````

So there are two bound issues:
1. `@method` annotation causes Psalm to blame wrong class.
2. MethodSignatureMismatch error is reported in a really inconsistent way. (It reports the issue something like two times per ten tries on my code with empty psalm cache, and doesn't report it at all in the sandbox: https://psalm.dev/r/dd7372a1bf No issues? Why?)

Also I don't know how to fix my `@psalm-method` annotation, because I can't specify array<string, mixed> inside a `@method`, it's a phpDoc syntax error, as I understand. At least looks like psalm doesn't understand this syntax, but doesn't yield a syntax error either. So any advice will be appreciated.

So, as I can't reproduce the issue(s) on psalm.dev properly as it tightly bound to the Doctrine library code, and I failed to find a minimal reproducible case, feel free to ask any debug info or point the place in the psalm's source code where I can investigate the issue(s).
